src/mds/MDCache.cc

```diff
@@ -12955,6 +12958,161 @@ void MDCache::enqueue_scrub_work(MDRequestRef& mdr)
+    mdcache->predirty_journal_parents(mdr, &le->metablob, in, nullptr,
+	PREDIRTY_PRIMARY);
+    mdcache->journal_dirty_inode(mdr.get(), &le->metablob, in);
+
+#if 0
+    mds->server->journal_and_reply(mdr, in, nullptr, le, onfinish);
+#else
+    mdr->committing = true;
+    std::string event_str("submit entry: ");
+    event_str += __func__;
+    mdr->mark_event(event_str);
+    dout(20) << "(uninline_data) submitting journal entry for " << *in << dendl;
+    mdlog->submit_entry(le, new C_MDC_DataUninlinedSubmitted(mdcache, in, mdr, onfinish));
+#endif
+  }
+};
+
+void MDCache::uninline_data_work(MDRequestRef& mdr)
+{
+  ceph_assert(mds->server->rdlock_path_pin_ref(mdr, true));
+  CInode *in = mdr->in[0];
+
+  MutationImpl::LockOpVec lov;
+  lov.add_xlock(&in->filelock);
+
+  if (!mds->locker->acquire_locks(mdr, lov)) {
+    dout(20) << "(uninline_data) acquire_locks failed; will retry later for " << *in << dendl;
+    return; // lock not available immediately
+  }
+
+  auto ino = [&]() { return in->ino(); };
+  auto pi = in->project_inode(mdr);
+  pi.inode->version = in->pre_dirty();
+  // pi.inode->change_attr++;
+  // pi.inode->xattr_version++;
+  auto mpi = in->_get_projected_inode();
+  auto objecter = mds->objecter;
+
+  ceph_assert(objecter);
+  ceph_assert(mpi->inline_data.version != CEPH_INLINE_NONE);
+
+  char oid_buf[32];
+  snprintf(oid_buf, sizeof(oid_buf), "%llx.00000000", (long long unsigned)ino().val);
+  object_t oid = oid_buf;
+
+  ObjectOperation create_ops;
+  create_ops.create(false);
+
+  dout(20) << "(uninline_data) dispatching objecter to create \""
+	   << mdr->get_filepath() << "\" for " << *in << dendl;
+
+  SnapContext snapc;
+  SnapRealm *snaprealm = in->find_snaprealm();
+  auto& snapc_ref = (snaprealm ? snaprealm->get_snap_context() : snapc);
+  objecter->mutate(oid,
+		   OSDMap::file_to_object_locator(mpi->layout),
+		   create_ops,
+		   snapc_ref,
+		   ceph::real_clock::now(),
+		   0,
+		   nullptr);
+
+  bufferlist inline_version_bl;
+  in->encode(inline_version_bl, mpi->inline_data.version);
+
+  ObjectOperation uninline_ops;
+  uninline_ops.cmpxattr("inline_version",
+			CEPH_OSD_CMPXATTR_OP_GT,
+			CEPH_OSD_CMPXATTR_MODE_U64,
+			inline_version_bl);
+  bufferlist inline_data;
+  mpi->inline_data.get_data(inline_data);
+  uninline_ops.write(0, inline_data, mpi->truncate_size, mpi->truncate_seq);
+  uninline_ops.setxattr("inline_version", std::to_string(CEPH_INLINE_NONE));
+
+  dout(20) << "(uninline_data) moving inline data for \"" << mdr->get_filepath() << "\" to file for " << *in << dendl;
+  C_SaferCond onfinish("MDCache::uninline_data_work wait for mutate");
+  objecter->mutate(oid,
+		   OSDMap::file_to_object_locator(mpi->layout),
+		   uninline_ops,
+		   snapc_ref,
+		   ceph::real_clock::now(),
+		   0,
+		   new C_IO_DataUninlined(in, this, mdr, onfinish));
+
+  dout(20) << "(uninline_data) mutation "
+	   << (onfinish.wait() == 0 ? "succeeded" : "failed") << dendl;
+}
```

Comment:
Maybe we should journal this. And do the uninline in the auth MDS only and notify all the replica MDSes at the same time ?